From b5f843fe7ece2b75a783fa785d6c7c6647a7a46d Mon Sep 17 00:00:00 2001
From: Michiiel <94533767+Michiiel@users.noreply.github.com>
Date: Thu, 30 May 2024 15:43:21 +0200
Subject: [PATCH] =?UTF-8?q?fix=20Hsync=20and=20Vsync=20polarity=20can't=20?=
 =?UTF-8?q?change=20from=20negatieve=20to=20positief=20=E2=80=A6=20(#6193)?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

vc4/hdmi: Fix Hsync and Vsync polarity changes

Polarity bits were only ever set and never cleared.
Make sure they can also be cleared.

Signed-off-by: Michiel Vanbiervliet <michiel.vanbiervliet@gmail.com>
---
 drivers/gpu/drm/vc4/vc4_hdmi.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/drivers/gpu/drm/vc4/vc4_hdmi.c b/drivers/gpu/drm/vc4/vc4_hdmi.c
index e42b431c29af..5ddd8d376b79 100644
--- a/drivers/gpu/drm/vc4/vc4_hdmi.c
+++ b/drivers/gpu/drm/vc4/vc4_hdmi.c
@@ -1883,14 +1883,15 @@ static void vc4_hdmi_encoder_post_crtc_enable(struct drm_encoder *encoder,
 	spin_lock_irqsave(&vc4_hdmi->hw_lock, flags);
 
 	HDMI_WRITE(HDMI_VID_CTL,
-		   HDMI_READ(HDMI_VID_CTL) |
-		   VC4_HD_VID_CTL_ENABLE |
-		   VC4_HD_VID_CTL_CLRRGB |
-		   VC4_HD_VID_CTL_UNDERFLOW_ENABLE |
-		   VC4_HD_VID_CTL_FRAME_COUNTER_RESET |
-		   VC4_HD_VID_CTL_BLANK_INSERT_EN |
-		   (vsync_pos ? 0 : VC4_HD_VID_CTL_VSYNC_LOW) |
-		   (hsync_pos ? 0 : VC4_HD_VID_CTL_HSYNC_LOW));
+	   (HDMI_READ(HDMI_VID_CTL) &~ 
+	   (VC4_HD_VID_CTL_VSYNC_LOW | VC4_HD_VID_CTL_HSYNC_LOW)) |
+	   VC4_HD_VID_CTL_ENABLE |
+	   VC4_HD_VID_CTL_CLRRGB |
+	   VC4_HD_VID_CTL_UNDERFLOW_ENABLE |
+	   VC4_HD_VID_CTL_FRAME_COUNTER_RESET |
+	   VC4_HD_VID_CTL_BLANK_INSERT_EN |
+	   (vsync_pos ? 0 : VC4_HD_VID_CTL_VSYNC_LOW) |
+	   (hsync_pos ? 0 : VC4_HD_VID_CTL_HSYNC_LOW));
 
 	HDMI_WRITE(HDMI_VID_CTL,
 		   HDMI_READ(HDMI_VID_CTL) & ~VC4_HD_VID_CTL_BLANKPIX);
-- 
2.34.1

